### PR TITLE
Allow pipeline to run in `bfloat16`

### PR DESCRIPTION
**Pull request by pcuenca**
Changes:
- Allow `dtype` to be specified on model load. This is a temporary solution until #567 is addressed in a more principled way.
- Convert loaded params to `bfloat16` or `float16` if necessary.
- Keep latents in `float32` during inference loop.

See comment: https://github.com/huggingface/diffusers/pull/559#issuecomment-1251877914

### CODE REVIEW

**Comment by patil-suraj** [comment]
The `dtype` here specifies the `dtype` of computation and should not be used for `params`. We should actually add a different argument to specify the dtype of parameters. For example, flax now has a `param_dtype` argument https://github.com/google/flax/blob/main/flax/linen/linear.py#L78.

The reason to do this is that, it's not recommended to keep `params` in `bfloat16` on TPUs during training as it can cause numerical instabilities, it's only done for extremely large models. Also having these two arguments allows you to do mixed-precision training: keep params in fp32 but do forward computation in half precision. 

For now we could create a branch with bfloat16 weights and the use that for inference.

**Comment by pcuenca** [comment]
I removed the automatic conversion but left the other change in order to be able to load the models specifying a `dtype`. @patil-suraj please, let me know if that's ok.

**Review by patil-suraj — APPROVED** [review]
LGTM, thanks a lot!

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py [hunk 1]
```diff
@@ -172,7 +172,7 @@ def __call__(
             self.unet.in_channels,
         )
         if latents is None:
-            latents = jax.random.normal(prng_seed, shape=latents_shape, dtype=self.dtype)
+            latents = jax.random.normal(prng_seed, shape=latents_shape, dtype=jnp.float32)
         else:
             if latents.shape != latents_shape:
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:177]: This file doesn't seem to be in `main`. Should be removed from here before mergin

Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:177]: The merge target is `flax_pipeline` for now.

**Comment by pcuenca** [comment]
Replaced by #600.

**Pull request closed without merging**